src/libraries/Common/src/System/Runtime/InteropServices/ComEventsMethod.cs

```diff
@@ -177,21 +180,28 @@ public void AddDelegate(Delegate d, bool wrapArgs = false)
                     }
                 }
 
-                var newWrapper = new DelegateWrapper(d, wrapArgs);
-                _delegateWrappers.Add(newWrapper);
-            }
+                newWrappers = wrappers.Count == 0
+                    ? new List<DelegateWrapper>()
+                    : wrappers.GetRange(0, wrappers.Count);
+
+                newWrappers.Add(new DelegateWrapper(d, wrapArgs));
+            } while (!PublishNewWrappers(newWrappers, wrappers));
         }
 
         public void RemoveDelegate(Delegate d, bool wrapArgs = false)
         {
-            lock (_delegateWrappers)
+            List<DelegateWrapper> wrappers;
+            List<DelegateWrapper> newWrappers;
+            do
             {
+                wrappers = _delegateWrappers;
+
                 // Find delegate wrapper index
                 int removeIdx = -1;
                 DelegateWrapper? wrapper = null;
-                for (int i = 0; i < _delegateWrappers.Count; i++)
+                for (int i = 0; i < wrappers.Count; i++)
                 {
-                    DelegateWrapper wrapperMaybe = _delegateWrappers[i];
+                    DelegateWrapper wrapperMaybe = wrappers[i];
                     if (wrapperMaybe.Delegate.GetType() == d.GetType() && wrapperMaybe.WrapArgs == wrapArgs)
                     {
                         removeIdx = i;
```

Comment:
Mainly I'm curious: are enough COM events subscribed to by multiple delegates in order to warrant this optimization of using Delegate.Combine rather than just having each delegate be its own entry in the list?  It looks safe from a cursory inspection, but it's a lot easier to prove correctness when the list being used is entirely immutable, whereas here the published list instance is having its contents modified.

Reply:
I think there is a bug when `Remove` is called on two threads in parallel:

First thread starts creates a new array with the delegate removed
Second thread removes the delegate in-place without creating a new array
First thread swaps in its array, reverting the affect of the second remove.

Reply:
> I think there is a bug when Remove is called on two threads in parallel:

Yeah, this is an example of the kind of issue that comes from trying to get sneaky (I hadn't noticed this particular case, but I believe you're right).

I think it'd be better to just do the simple thing and actually have the list be immutable; then the only concurrency management need be the simple act of publishing safely.

Or if our goal is to stick with what netfx had, just revert the locking entirely and go back to exactly what netfx had :)